### PR TITLE
feat: Add `TxEip2930Gadget` to handle EIP-2930 (for evm-circuit part)

### PR DESCRIPTION
**Pull request by silathdiir**
### Summary

In bus-mapping:

- add two copy events for access list addresses and storage keys to begin-tx (will handle in copy-circuit for another PR).

In zkevm-circuits:

- add `AccessListAddressesLen` and `AccessListStorageKeysLen` to tx-table.
- add `TxEip2930Gadget` to begin-tx for copy-circuit lookups (and move `AccessListGasCost` into this gadget).
- move `tx_type` from `TxL1MsgGadget` to `BeginTxGadget`, since it's used for both TxL1MsgGadget and TxEip2930Gadget.

### Reference

https://www.notion.so/scrollzkp/RLP-Tx-Copy-EVM-circuit-constrain-1559-tx-58a7ed60c0354b97877472df0c35f15a

### Asana

https://app.asana.com/0/1202392035220922/1206015154508065/f
https://app.asana.com/0/1202392035220922/1206015154508066/f

### TODO

Will update copy circuit for two copy events in another PR.

### CODE REVIEW

**Comment by silathdiir** [comment]
\b run testool

File: bus-mapping/src/circuit_input_builder/execution.rs [hunk 1]
```diff
@@ -221,6 +221,14 @@ pub enum CopyDataType {
...
     /// This is used for Copy Lookup from SHA3 opcode verification.
     RlcAcc,
+    /// When the source for copy event is access-list addresses (EIP-2930) of tx-table.
+    TxAccessListAddresses,
+    /// When the destination for copy event is access-list addresses (EIP-2930) of rw-table.
+    RwAccessListAddresses,
+    /// When the source for copy event is access-list storage keys (EIP-2930) of tx-table.
+    TxAccessListStorageKeys,
+    /// When the destination for copy event is access-list storage keys (EIP-2930) of rw-table.
+    RwAccessListStorageKeys,
```
Comment (roynalnaruto) [inline bus-mapping/src/circuit_input_builder/execution.rs]: I don't think this is the right approach to add 4 new enum variants. The copy circuit uses a `BinaryNumberGadget` to check for the data type. And we will need more bits to support these new types, hence will also increase the degree of the circuit.

Comment (roynalnaruto) [reply bus-mapping/src/circuit_input_builder/execution.rs]: Since the `(src, dst)` tuples for these 2 cases are unique, i.e. `(TxAccessListAddresses, RwAccessListAddresses)` and `(TxAcessListStorageKeys, RwAccessListStorageKeys)` are the only 2 tuples. This means instead of 4 variants, we can also do with 2 variants and use the variant for both `source` and `destination`.

This would mean we will continue using the same number of bits in the copy circuit for the data type `BinaryNumberGadget`.

File: bus-mapping/src/evm/opcodes/begin_end_tx.rs [hunk 2]
```diff
@@ -615,3 +618,57 @@ fn gen_tx_l1_fee_ops(
...
+
+    // Add copy event for access-list storage keys.
+    let rw_counter_start = state.block_ctx.rwc;
+    state.push_copy(
+        exec_step,
+        CopyEvent {
+            src_addr: 0,
+            src_addr_end: storage_key_size,
+            src_type: CopyDataType::TxAccessListStorageKeys,
+            src_id: tx_id.clone(),
```
Comment (roynalnaruto) [inline bus-mapping/src/evm/opcodes/begin_end_tx.rs:661]: you may not need this clone here. Do you?

File: bus-mapping/src/evm/opcodes/begin_end_tx.rs [hunk 3]
```diff
@@ -615,3 +618,57 @@ fn gen_tx_l1_fee_ops(
...
+            src_addr_end: address_size,
+            src_type: CopyDataType::TxAccessListAddresses,
+            src_id: tx_id.clone(),
+            dst_addr: 0,
+            dst_type: CopyDataType::RwAccessListAddresses,
+            dst_id: tx_id.clone(),
+            log_id: None,
+            rw_counter_start,
+            // TODO
+            copy_bytes: CopyBytes::new(vec![], None, None),
```
Comment (roynalnaruto) [inline bus-mapping/src/evm/opcodes/begin_end_tx.rs:649]: Is this the part that will be done in another PR?

File: eth-types/src/geth_types.rs [hunk 4]
```diff
@@ -49,12 +49,22 @@ impl From<TxType> for u64 {
...
+    /// If this type is EIP-155 or not
     pub fn is_eip155_tx(&self) -> bool {
-        matches!(*self, TxType::Eip155)
+        matches!(*self, Self::Eip155)
+    }
+
+    /// If this type is EIP-1559 or not
+    pub fn is_eip1559_tx(&self) -> bool {
+        matches!(*self, Self::Eip1559)
+    }
```
Comment (roynalnaruto) [inline eth-types/src/geth_types.rs]: If this PR is handling only EIP-2930 related changes, better not include this here, so we can account for it in the EIP-1559 PR.

File: zkevm-circuits/src/table.rs [hunk 5]
```diff
@@ -177,6 +175,12 @@ pub enum TxFieldTag {
     TxHash,
     /// TxType: Type of the transaction
     TxType,
+    /// Access list address count (EIP 2930)
+    AccessListAddressesLen,
+    /// Access list all storage key count (EIP 2930)
+    AccessListStorageKeysLen,
+    /// Gas cost for access list (EIP 2930)
+    AccessListGasCost,
```
Comment (roynalnaruto) [inline zkevm-circuits/src/table.rs]: We will also need a `AccessListRlc` field type. However I am not sure if @darth-cy is adding it as a part of his PR. If he is not, it should be added here.

File: zkevm-circuits/src/tx_circuit.rs [hunk 6]
```diff
@@ -42,7 +43,7 @@ use crate::{
 use bus_mapping::circuit_input_builder::keccak_inputs_sign_verify;
 use eth_types::{
     geth_types::{
-        TxType,
+        access_list_address_and_storage_key_sizes, TxType,
```
Comment (roynalnaruto) [inline zkevm-circuits/src/tx_circuit.rs]: I think it will make sense to rename this to: `access_list_size` and the docs on the method can explain what the tuple return value stands for, instead of having such a long function name.

File: eth-types/src/geth_types.rs [hunk 7]
```diff
@@ -402,3 +412,18 @@ impl GethData {
     }
 }
 */
+
+/// Return access list address size and all storage key size.
+pub fn access_list_address_and_storage_key_sizes(access_list: &Option<AccessList>) -> (u64, u64) {
```
Comment (roynalnaruto) [inline eth-types/src/geth_types.rs]:
```suggestion
/// Returns the number of addresses and the cumulative number of storage keys in the entire access list.
pub fn access_list_size(access_list: &Option<AccessList>) -> (u64, u64) {
```

File: bus-mapping/src/evm/opcodes/begin_end_tx.rs [hunk 8]
```diff
@@ -615,3 +618,57 @@ fn gen_tx_l1_fee_ops(
...
+        return Ok(());
+    }
+
+    let tx_id = NumberOrHash::Number(state.tx_ctx.id());
+    let (address_size, storage_key_size) =
+        access_list_address_and_storage_key_sizes(&state.tx.access_list);
+
+    // Add copy event for access-list addresses.
+    let rw_counter_start = state.block_ctx.rwc;
+    state.push_copy(
```
Comment (roynalnaruto) [inline bus-mapping/src/evm/opcodes/begin_end_tx.rs:636]: Do you also have a copy event if the access list is empty, i.e. `address_size == 0 && storage_key_size == 0` ?

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            .condition(is_eip2930_tx.expr(), |cb| {
+                let [address_len, storage_key_len, gas_cost] = [
+                    TxFieldTag::AccessListAddressesLen,
+                    TxFieldTag::AccessListStorageKeysLen,
+                    TxFieldTag::AccessListGasCost,
+                ]
+                .map(|field_tag| cb.tx_context(tx_id.expr(), field_tag, None));
+
+                // Let copy-circuit to write the tx-table's access list addresses into rw-table.
+                cb.copy_table_lookup(
```
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs:46]: Comment related to the bus-mapping `state.push_copy`. If you push the event conditionally (whether the access list is non-empty), then you will need an additional condition here so that the lookup does not fail.

**Review by roynalnaruto — CHANGES_REQUESTED** [review]
As I pointed out, we can reduce the number of variants in the `CopyDataType` since there are only 2 unique `(src, dst)` tuples.

Other than that, I have raised some minor points, refactoring points. And a few questions. Looks good otherwise :)

File: bus-mapping/src/evm/opcodes/begin_end_tx.rs (hunk 2 above)
Comment (silathdiir) [reply bus-mapping/src/evm/opcodes/begin_end_tx.rs:661]: Seems it's needed, since it's constructed as [NumberOrHash::Number here](https://github.com/scroll-tech/zkevm-circuits/pull/1045/files/10436947f181c545f5f3950f49be0b1386ab6d8a#diff-d8df2606a9422d81101a23b7e700280608c770cc3c98b6f214dd8e207835efc3R631) and also used as [dst_id here](https://github.com/scroll-tech/zkevm-circuits/pull/1045/files/10436947f181c545f5f3950f49be0b1386ab6d8a#diff-d8df2606a9422d81101a23b7e700280608c770cc3c98b6f214dd8e207835efc3R665).

File: bus-mapping/src/evm/opcodes/begin_end_tx.rs (hunk 3 above)
Comment (silathdiir) [reply bus-mapping/src/evm/opcodes/begin_end_tx.rs:649]: Yes, suppose to add in the following PR which adds the related code of copy-circuit.

File: bus-mapping/src/circuit_input_builder/execution.rs (hunk 1 above)
Comment (silathdiir) [reply bus-mapping/src/circuit_input_builder/execution.rs]: Merge `CopyDataType` of access list to `AccessListAddresses` and `AccessListStorageKeys` (for both source and destination) in commit 097d3b8.

File: eth-types/src/geth_types.rs (hunk 4 above)
Comment (silathdiir) [reply eth-types/src/geth_types.rs]: Delete this `is_eip1559_tx` function in commit [cecb6c1](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/cecb6c1c1a85b77f2d755980de9df4a39511a080).

File: zkevm-circuits/src/tx_circuit.rs (hunk 6 above)
Comment (silathdiir) [reply zkevm-circuits/src/tx_circuit.rs]: Rename function `access_list_address_and_storage_key_sizes` to `access_list_size` in commit [1988777](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/198877743211fbaef748891523986f5ace1c3293).

File: eth-types/src/geth_types.rs (hunk 7 above)
Comment (silathdiir) [reply eth-types/src/geth_types.rs]: Rename the function and comment with other used places in commit [1988777](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/198877743211fbaef748891523986f5ace1c3293), thanks.

File: bus-mapping/src/evm/opcodes/begin_end_tx.rs (hunk 8 above)
Comment (silathdiir) [reply bus-mapping/src/evm/opcodes/begin_end_tx.rs:636]: Yes, suppose to also push copy events even if `address_size` or `storage_key_size` is zero to make consistent (if no big effect), only check if it's EIP-2930 tx in [bus-mapping](https://github.com/scroll-tech/zkevm-circuits/pull/1045/files/10436947f181c545f5f3950f49be0b1386ab6d8a#diff-d8df2606a9422d81101a23b7e700280608c770cc3c98b6f214dd8e207835efc3R627) and [zkevm-circuits](https://github.com/scroll-tech/zkevm-circuits/pull/1045/files/10436947f181c545f5f3950f49be0b1386ab6d8a#diff-2244d79548b85a69ee99b1f17962710d46fc454599c3cd5920a2995e74e39033R38).

File: zkevm-circuits/src/table.rs (hunk 5 above)
Comment (silathdiir) [reply zkevm-circuits/src/table.rs]: Add `AccessListRLC` in commit cd6ac3a. May need @darth-cy help check if [the RLC calculation](https://github.com/scroll-tech/zkevm-circuits/pull/1045/files#diff-864796c9d6f832182c9e2c72268681463f4af9eaac5cf56b04f56b3a614923fbR1972-R1980) is correct when implements constraints in tx-circuit.

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs (hunk 9 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs:46]: I suppose if could push copy events even if address_size or storage_key_size is zero to make consistent (if no big effect?). WDYT? Thanks.

File: zkevm-circuits/src/table.rs [hunk 10]
```diff
@@ -177,6 +175,14 @@ pub enum TxFieldTag {
     TxHash,
     /// TxType: Type of the transaction
     TxType,
+    /// Access list address count (EIP-2930)
+    AccessListAddressesLen,
+    /// Access list all storage key count (EIP-2930)
+    AccessListStorageKeysLen,
+    /// Gas cost for access list (EIP-2930)
+    AccessListGasCost,
```
Comment (roynalnaruto) [inline zkevm-circuits/src/table.rs]: We can actually remove this field `AccessListGasCost` and simply use the two above fields, i.e. `AccessListAddressesLen` and `AccessListStorageKeysLen` to calculate the gas cost in the `TxEip2930Gadget`.

File: zkevm-circuits/src/tx_circuit.rs [hunk 11]
```diff
@@ -346,6 +346,10 @@ impl<F: Field> SubCircuitConfig<F> for TxCircuitConfig<F> {
         is_tx_tag!(is_hash, TxHash);
         is_tx_tag!(is_block_num, BlockNumber);
         is_tx_tag!(is_tx_type, TxType);
+        is_tx_tag!(is_access_list_addresses_len, AccessListAddressesLen);
+        is_tx_tag!(is_access_list_storage_keys_len, AccessListStorageKeysLen);
+        is_tx_tag!(is_access_list_gas_cost, AccessListGasCost);
```
Comment (roynalnaruto) [inline zkevm-circuits/src/tx_circuit.rs]: If you remove that field, you won't need this.

File: zkevm-circuits/src/tx_circuit.rs [hunk 12]
```diff
@@ -79,11 +80,11 @@ use halo2_proofs::plonk::SecondPhase;
 use itertools::Itertools;
 
 /// Number of rows of one tx occupies in the fixed part of tx table
-pub const TX_LEN: usize = 24;
+pub const TX_LEN: usize = 27;
```
Comment (roynalnaruto) [inline zkevm-circuits/src/tx_circuit.rs]: This will also change if you remove that field `AccessListGasCost`

File: zkevm-circuits/src/tx_circuit.rs [hunk 13]
```diff
@@ -467,13 +471,16 @@ impl<F: Field> SubCircuitConfig<F> for TxCircuitConfig<F> {
...
-                (is_access_list_gas_cost(meta), Null),
                 (is_sign_hash(meta), Null),
                 (is_hash(meta), Null),
                 (is_data(meta), Null),
                 (is_block_num(meta), Null),
                 (is_chain_id_expr(meta), Tag::ChainId.into()),
                 (is_tx_type(meta), Null),
+                (is_access_list_addresses_len(meta), Null),
+                (is_access_list_storage_keys_len(meta), Null),
+                (is_access_list_gas_cost(meta), Null),
```
Comment (roynalnaruto) [inline zkevm-circuits/src/tx_circuit.rs]: This will be removed

File: zkevm-circuits/src/tx_circuit.rs [hunk 14]
```diff
@@ -559,7 +566,8 @@ impl<F: Field> SubCircuitConfig<F> for TxCircuitConfig<F> {
             cb.gate(meta.query_fixed(q_enable, Rotation::cur()))
         });
 
-        // TODO: add constraints for AccessListGasCost.
+        // TODO: add constraints for AccessListAddressesLen, AccessListStorageKeysLen,
+        // AccessListGasCost and AccessListRLC.
```
Comment (roynalnaruto) [inline zkevm-circuits/src/tx_circuit.rs]:
```suggestion
        // TODO: add constraints for AccessListAddressesLen, AccessListStorageKeysLen
        // and AccessListRLC.
```

File: zkevm-circuits/src/tx_circuit.rs [hunk 15]
```diff
@@ -1941,6 +1946,38 @@ impl<F: Field> TxCircuitConfig<F> {
...
+            (
+                AccessListStorageKeysLen,
+                None,
+                Value::known(F::from(access_list_storage_key_size)),
+            ),
+            (
+                AccessListGasCost,
+                None,
+                Value::known(F::from(tx.access_list_gas_cost)),
+            ),
```
Comment (roynalnaruto) [inline zkevm-circuits/src/tx_circuit.rs]: This will not be needed

File: zkevm-circuits/src/witness/tx.rs [hunk 16]
```diff
@@ -306,6 +308,38 @@ impl Transaction {
...
+                Value::known(F::from(self.id as u64)),
+                Value::known(F::from(TxContextFieldTag::AccessListStorageKeysLen as u64)),
+                Value::known(F::zero()),
+                Value::known(F::from(access_list_storage_key_size)),
+            ],
+            [
+                Value::known(F::from(self.id as u64)),
+                Value::known(F::from(TxContextFieldTag::AccessListGasCost as u64)),
+                Value::known(F::zero()),
+                rlc_be_bytes(&self.rlp_signed, challenges.keccak_input()),
```
Comment (roynalnaruto) [inline zkevm-circuits/src/witness/tx.rs]: Is this correct?

File: zkevm-circuits/src/witness/tx.rs [hunk 17]
```diff
@@ -306,6 +308,38 @@ impl Transaction {
...
+            ],
```
Comment (roynalnaruto) [inline zkevm-circuits/src/witness/tx.rs]: But you won't need this block if you remove the `AccessListGasCost` field

File: zkevm-circuits/src/witness/tx.rs [hunk 18]
```diff
@@ -866,7 +901,7 @@ impl From<MockTransaction> for Transaction {
             call_data: mock_tx.input.to_vec(),
             call_data_length: mock_tx.input.len(),
             call_data_gas_cost: tx_data_gas_cost(&mock_tx.input),
-            access_list_gas_cost: tx_access_list_gas_cost(&Some(mock_tx.access_list)),
+            access_list_gas_cost: tx_access_list_gas_cost(&access_list),
```
Comment (roynalnaruto) [inline zkevm-circuits/src/witness/tx.rs]: you won't need this as well then

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+    plonk::{Error, Expression},
+};
+
+/// Transaction EIP-2930 gadget to handle optional access-list
+#[derive(Clone, Debug)]
+pub(crate) struct TxEip2930Gadget<F> {
+    is_eip2930_tx: IsEqualGadget<F>,
+    access_list_address_len: Cell<F>,
+    access_list_storage_key_len: Cell<F>,
+    access_list_gas_cost: Cell<F>,
```
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs]: you won't need this field

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+        tx_type: Expression<F>,
+    ) -> Self {
+        let is_eip2930_tx = IsEqualGadget::construct(cb, tx_type, (TxType::Eip2930 as u64).expr());
+
+        let [access_list_address_len, access_list_storage_key_len, access_list_gas_cost] = cb
+            .condition(is_eip2930_tx.expr(), |cb| {
+                let [address_len, storage_key_len, gas_cost] = [
+                    TxFieldTag::AccessListAddressesLen,
+                    TxFieldTag::AccessListStorageKeysLen,
+                    TxFieldTag::AccessListGasCost,
```
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs]: you won't need this

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+                );
+
+                [address_len, storage_key_len, gas_cost]
+            });
+
+        Self {
+            is_eip2930_tx,
+            access_list_address_len,
+            access_list_storage_key_len,
+            access_list_gas_cost,
```
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs]: you won't need this

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+
+        Ok(())
+    }
+
+    pub(crate) fn gas_cost(&self) -> Expression<F> {
+        select::expr(
+            self.is_eip2930_tx.expr(),
+            self.access_list_gas_cost.expr(),
+            0.expr(),
+        )
```
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs]:
```suggestion
        select::expr(
            self.is_eip2930_tx.expr(),
            self.access_list_address_len.expr() * 1900.expr()
                + self.access_list_storage_key_len.expr() * 2400.expr(),
            0.expr(),
        )
```

However, please use the `1900` and `2400` from `eth_types::evm_types::GasCost`

**Review by roynalnaruto — CHANGES_REQUESTED** [review]
Did another round. I think we can remove the `AccessListGasCost` field from the `TxTable` completely and instead simply make the `TxEip2930Gadget` return the correctly computed gas cost for the access list. This removes a field that we don't really need. Otherwise looks good!

File: zkevm-circuits/src/tx_circuit.rs (hunk 11 above)
Comment (silathdiir) [reply zkevm-circuits/src/tx_circuit.rs]: Removed in commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83).

File: zkevm-circuits/src/tx_circuit.rs (hunk 12 above)
Comment (silathdiir) [reply zkevm-circuits/src/tx_circuit.rs]: Decrease 1 (to `26`) in commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83).

File: zkevm-circuits/src/tx_circuit.rs (hunk 13 above)
Comment (silathdiir) [reply zkevm-circuits/src/tx_circuit.rs]: Removed in commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83).

File: zkevm-circuits/src/tx_circuit.rs (hunk 14 above)
Comment (silathdiir) [reply zkevm-circuits/src/tx_circuit.rs]: Update this comment in the same commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83).

File: zkevm-circuits/src/tx_circuit.rs (hunk 15 above)
Comment (silathdiir) [reply zkevm-circuits/src/tx_circuit.rs]: Removed in commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83).

File: zkevm-circuits/src/witness/tx.rs (hunk 16 above)
Comment (silathdiir) [reply zkevm-circuits/src/witness/tx.rs]: Sorry, it's wrong copy. Just remove this with `AccessListGasCost` in commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83).

File: zkevm-circuits/src/witness/tx.rs (hunk 17 above)
Comment (silathdiir) [reply zkevm-circuits/src/witness/tx.rs]: Yes.

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs (hunk 19 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs]: Yes, removed in the same commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83).

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs (hunk 20 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs]: Yes, removed it in commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83), and calculate gas-cost as:
```
access_list_address_len.expr() * GasCost::ACCESS_LIST_PER_ADDRESS.expr()
    + access_list_storage_key_len.expr() * GasCost::ACCESS_LIST_PER_STORAGE_KEY.expr()
```

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs (hunk 22 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs]: Yes, fixed it in commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83):
```
access_list_address_len.expr() * GasCost::ACCESS_LIST_PER_ADDRESS.expr()
    + access_list_storage_key_len.expr() * GasCost::ACCESS_LIST_PER_STORAGE_KEY.expr()
```

File: zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs (hunk 21 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs]: Removed in commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83).

File: zkevm-circuits/src/table.rs (hunk 10 above)
Comment (silathdiir) [reply zkevm-circuits/src/table.rs]: Removed in commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83).

File: zkevm-circuits/src/witness/tx.rs (hunk 18 above)
Comment (silathdiir) [reply zkevm-circuits/src/witness/tx.rs]: Removed in the same commit [6880378](https://github.com/scroll-tech/zkevm-circuits/pull/1045/commits/68803783b1447b8fefe2c1fabd626beae51b8a83).

**Review by roynalnaruto — APPROVED** [review]
Looks good! 👍
